examples/magento-graphcms/components/GraphCMS/pageContent.ts

```diff
@@ -0,0 +1,176 @@
+   * in stale data.
+   *
+   * Todo: Implement next.js 13 fetch revalidation:
+   * https://beta.nextjs.org/docs/data-fetching/fetching#revalidating-data
+   */
+  const alwaysCache = process.env.NODE_ENV !== 'development' ? 'cache-first' : undefined
+  const fetchPolicy = cached ? alwaysCache : undefined
+
+  const allRoutes = await client.query({ query: AllPageRoutesDocument, fetchPolicy: alwaysCache })
+
+  // Only do the query when there the page is found in the allRoutes
+  const found = allRoutes.data.pages.some((page) => page.url === url)
+
+  return found
+    ? client.query({ query: DefaultPageDocument, variables: { url }, fetchPolicy })
+    : Promise.resolve({ data: { pages: [] } })
+}
+
+/**
+ * - Boven de product description zetten? in rowrenderer zetten
+ * - Hoe gaan we dit optioneel maken?
+ * - Hoe gaan we dit upgradebaar maken? management sdk
+ */
+
+/**
+ * Fetch the page content for the given urls.
+ *
+ * - Uses an early bailout to check to reduce hygraph calls.
+ * - Implements an alias sytem to merge the content of multiple pages.
+ */
+export async function hygraphDynamicContent(
+  client: ApolloClient<NormalizedCacheObject>,
+  pageQuery: Promise<{ data: DefaultPageQuery }>,
+  propertiesPromise: Promise<object> | object,
+  cached = false,
+): Promise<{ data: DefaultPageQuery }> {
+  const alwaysCache = process.env.NODE_ENV !== 'development' ? 'cache-first' : undefined
+  const fetchPolicy = cached ? alwaysCache : undefined
+
+  const allRoutes = await client.query({ query: AllPageRoutesDocument, fetchPolicy: alwaysCache })
+
+  // Get the required rowIds from the conditions
+  const properties = await propertiesPromise
+  const rowIds = allRoutes.data.dynamicRows
+    .filter((availableDynamicRow) =>
+      availableDynamicRow.conditions.some((condition) => matchCondition(condition, properties)),
+    )
+    .map((row) => row.id)
+
+  const dynamicRows =
+    rowIds.length !== 0
+      ? client.query({ query: DynamicRowDocument, variables: { rowIds }, fetchPolicy })
+      : undefined
+
+  const [pageResult, dynamicResult] = await Promise.all([pageQuery, dynamicRows])
+
+  // Create a copy of the content array.
+  const content = [...(pageResult.data.pages[0]?.content ?? [])]
+
+  dynamicResult?.data.dynamicRows.forEach((dynamicRow) => {
+    const { placement, target, row } = dynamicRow
+    if (!row) return
+    if (!target) {
+      if (placement === 'BEFORE') content.unshift(row)
+      else content.push(row)
+      return
+    }
+
+    const targetIdx = content.findIndex((c) => c.id === target.id)
+    if (placement === 'BEFORE') content.splice(targetIdx, 0, row)
+    if (placement === 'AFTER') content.splice(targetIdx + 1, 0, row)
+    if (placement === 'REPLACE') content.splice(targetIdx, 1, row)
+  })
+
+  // Return the merged page result.
+  return { data: { ...pageResult.data, pages: [{ ...pageResult.data.pages[0], content }] } }
+}
```

Comment:
This can be moved to the graphcms-ui package since we can assume pages and dynamicRows exist.